internal/k8s/reconciler/route.go

```diff
@@ -362,8 +363,15 @@ func (r *K8sRoute) Validate(ctx context.Context) error {
 		for _, httpRule := range route.Spec.Rules {
 			rule := httpRule
 			routeRule := service.NewRouteRule(&rule)
+
 			for _, backendRef := range rule.BackendRefs {
 				ref := backendRef
+
+				allowed, err := routeAllowedForHTTPBackend(ctx, route, ref, r.client)
+				if err != nil || !allowed {
+					return fmt.Errorf("route not allowed for backend %q", ref.Name)
```

Comment:
I know it's reserved for follow-up, but I'd actually probably prefer just no-op'ing in this branch if we're not going to set the `Status` in this commit, namely because this is going to trigger the reconciliation loop to re-run by throwing an error here.

RE statuses, it's not too bad what you'll need to do -- basically just add an entry for `RefNotPermitted` as per [this part of the spec](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1alpha2.HTTPBackendRef) to `statuses.yaml` here:

https://github.com/hashicorp/consul-api-gateway/blob/8ff752ad0295df4f3daee27b5c67b82f42b75694/internal/k8s/reconciler/config/statuses.yaml#L129

Regenerate the status structures with `go generate ./...` and you should be able to do something like `status.ResolvedRefs.RefNoPermitted = fmt.Errorf("route not allowed for backend %q", ref.Name)`

The one caveat to this and what makes status values in the API spec really wonky (and we may want to actually raise upstream) is that the spec is inconsistent AFAICT with regard to the particularities of `ResolvedRefs`, specifically, from the link above:

> If there is a cross-namespace reference to an existing object that is not covered by a ReferencePolicy, the controller must ensure the “ResolvedRefs” condition on the Route is set to status: False, with the “RefNotPermitted” reason and not configure this backend in the underlying implementation.

While later on there's also a value for `RefNotPermitted` in the listener status:

> This reason is used with the “ResolvedRefs” condition when one of the Listener’s Routes has a BackendRef to an object in another namespace, where the object in the other namespace does not have a ReferencePolicy explicitly allowing the reference.

So interestingly enough, there at least seems to me to be uncertainty about whether a route should be considered "bound" or not to a listener if it's missing a `ReferencePolicy` and whether or not the whole listener should then have a bad status (i.e. is this a condition for the route or for the listener?). I'm assuming that the latter is a mistake in the spec and that this properly belongs on the route and should just prevent the route from binding to the listener. Yay the complexities of statuses!

Reply:
logging for now instead

Reply:
Opened an issue upstream at https://github.com/kubernetes-sigs/gateway-api/issues/1112 about the lack of clarity in how Listener "acceptance" of a route and corresponding Route and Listener status conditions should be set.